From 4fb6a2ab41bb9b8623e50b52339e35dcf1f0619f Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:04:43 -0400
Subject: [x86] perf/intel: Use proper dTLB-load-misses event on IvyBridge

Message-id: <1410181752-25631-39-git-send-email-jolsa@redhat.com>
Patchwork-id: 91713
O-Subject: [PATCH RHEL7.1 BZ1134356 038/307] perf/x86/intel: Use proper dTLB-load-misses event on IvyBridge
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 1996388e9f4e3444db8273bc08d25164d2967c21
Author: Vince Weaver <vincent.weaver@maine.edu>
Date: Mon Jul 14 15:33:25 2014 -0400

description
===========
This was discussed back in February:

 https://lkml.org/lkml/2014/2/18/956

But I never saw a patch come out of it.

On IvyBridge we share the SandyBridge cache event tables, but the
dTLB-load-miss event is not compatible.  Patch it up after
the fact to the proper DTLB_LOAD_MISSES.DEMAND_LD_MISS_CAUSES_A_WALK
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/cpu/perf_event_intel.c b/arch/x86/kernel/cpu/perf_event_intel.c
index 33403b3..b5bd37d 100644
--- a/arch/x86/kernel/cpu/perf_event_intel.c
+++ b/arch/x86/kernel/cpu/perf_event_intel.c
@@ -2470,6 +2470,9 @@ __init int intel_pmu_init(void)
  case 62: /* IvyBridge EP */
   memcpy(hw_cache_event_ids, snb_hw_cache_event_ids,
          sizeof(hw_cache_event_ids));
+  /* dTLB-load-misses on IVB is different than SNB */
+  hw_cache_event_ids[C(DTLB)][C(OP_READ)][C(RESULT_MISS)] = 0x8108; /* DTLB_LOAD_MISSES.DEMAND_LD_MISS_CAUSES_A_WALK */
+
   memcpy(hw_cache_extra_regs, snb_hw_cache_extra_regs,
          sizeof(hw_cache_extra_regs));
 
-- 
1.7.1